javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+#### Extra credit
+
+1. You may have exposed the list that the piggy bank uses to track coins as a public property. Depending on how you implemented the piggy bank, modifying this list without using something like a `deposit` or `withdraw` method on the object could throw some of it's properties out of wack, like the one that tells you the amount of savings you've accrued.
+
+   Additionally, adding an unexpected value to this list by modifying it directly could cause errors within your piggy bank methods. Let's try to ensure that doesn't happen!
+   - Indicate that the list of coins the piggy bank uses is a **private property** of the object.
+   - Create a **public method** that gives the user a *copy* of the list that they can manipulate to their hearts content without breaking the piggy bank object.
+
+1. Try creating a couple 'coin' objects to deposit into the piggy bank instead of using a string.
+   - A coin object might contain properties like the name of the coin, and it's value (preferably in the same unit the piggy bank measures it's savings).
+   - Make sure that your piggy bank object is still able to keep track of savings and remove coins correctly after this change!
+   - After making this change, consider how you are calculating savings for the piggy bank compared to before. Do you prefer working with the objects or the strings more?
+   - Consider how the piggy bank might check for a correct string vs checking for an object with the correct properties when depositing. Which seems easier? Does one seem more flexible than the other?
+
+</div>
+
+### Knowledge check
+
+The following questions are an opportunity to reflect on key topics in this lesson. If you can't answer a question, click on it to review the material, but keep in mind you are not expected to memorize or master this knowledge.
+
+- [What are two ways you can use objects to organize code?](#objects-as-a-data-structure)
+- [What is a 'method'?](#objects-as-a-design-pattern)
+- [What is the `this` keyword used for?](#objects-as-a-design-pattern)
+- [What methods should exist as part of an object's public interface?](#objects-as-a-design-pattern)
+
+### Additional resources
+
+This section contains helpful links to related content. It isn't required, so consider it supplemental.
+
+- [The Principles of Object-Oriented JavaScript](https://www.amazon.com/Principles-Object-Oriented-JavaScript-Nicholas-Zakas/dp/1593275404) book by
+Nicholas C. Zakas is really great to understand OOP in JavaScript, which explains concepts in-depth, which explores JavaScript's object-oriented nature, revealing the language's unique implementation of inheritance and other key characteristics, it's not free but it's very valuable.
```

Comment:
No manual line break needed.
I'm not 100% sure this book would be an additional resource in scope. Related? Yes. Valuable for some people to dig into now before having had experience with many other things? I feel it's still out of scope and a rabbit hole for many people. Personally, I wouldn't include this resource.
```suggestion
- [The Principles of Object-Oriented JavaScript](https://www.amazon.com/Principles-Object-Oriented-JavaScript-Nicholas-Zakas/dp/1593275404) book by Nicholas C. Zakas is really great to understand OOP in JavaScript, which explains concepts in-depth, which explores JavaScript's object-oriented nature, revealing the language's unique implementation of inheritance and other key characteristics, it's not free but it's very valuable.
```

Reply:
I believe this was in the original Objects and Object Constructors lesson at some point, but may have been removed.

It might be worth tucking it into the Object Constructors lesson here. I gave it a *very* quick read a while ago after doing the JS course, and while I didn't find it valuable personally, I think for the learner going over this stuff for the first time it could be appropriate and give some insight. It may be easier for them to make sense of now too, after adding the "Organizing Code with Objects" lesson.

Since it's not particularly updated, it uses object constructors primarily to explore classical inheritance a little bit (but not particularly in-depth). It may be worth noting that it's a bit dated, if it's to be added back.

Let me know what you think about moving it back into the Object Constructors lesson, or just wiping it out all together.

Reply:
The thing with linking books is that they often go way further than they need to for this part of the curriculum, and are places for huge rabbit holes. More often than not, only certain parts of the book are relevant. At this point in the curriculum, especially for those in the JS pathway, people will still be fairly new to many of these concepts. Books tend to have more value for those who have a little more experience where exploring out of scope provides more benefits than rabbit hole opportunities.

Relevant discord message: https://discord.com/channels/505093832157691914/505093832157691916/1296589566530293793

Also note that the team have been considering rethinking our approach to additional resources anyway, as many lessons have ARs that go way out of scope than what is appropriate for the lesson and curriculum position, and/or have way too many ARs that are just "things people find helpful" (big hole for anyone to throw any old resource in because everyone will find a different wording nicer than others). So I'm also coming at this from a more cautious and smaller-scoped angle.